networked_controller.cpp

```diff
@@ -1530,211 +1433,140 @@ DollController::DollController(NetworkedController *p_node) :
+	const int64_t current_virtual_delay = int64_t(future_epoch) - int64_t(current_epoch);
 
-		batch_young_epoch = MIN(epoch, batch_young_epoch);
+	if (current_epoch > future_epoch) {
+		// Make sure we set back the current epoch in case of overshot.
+		// The overshot is wanted to correctly calculate `current_virtual_delay`, but at this point
+		// the overshot need to be normalized.
+		current_epoch = future_epoch;
 	}
+	past_epoch = current_epoch;
+	future_epoch = epoch;
 
-	// ~~ Establish the interpolation speed ~~
-	if (batch_young_epoch == UINT32_MAX) {
-		// This may just be a late arrived batch, so nothing more to do.
-		return;
-	}
+	// Make sure to store the current state, so we can use it to interpolate.
+	// This is necessary so we allow a bit of overshot.
+	past_epoch_buffer.begin_write(0);
+	node->native_collect_epoch_data(past_epoch_buffer);
 
-	const real_t net_sentitivity = node->get_doll_net_sensitivity();
+	// ~~ Establish the interpolation speed ~~
 
 	// Establish the connection quality by checking if the batch takes
 	// always the same time to arrive.
 	const uint32_t now = OS::get_singleton()->get_ticks_msec();
+
 	// If now is bigger, then the timer has been disabled, so we assume 0.
-	network_watcher.push(now > batch_receiver_timer ? now - batch_receiver_timer : 0);
-	batch_receiver_timer = now;
 
-	const uint32_t avg_receive_time = network_watcher.average();
-	const real_t deviation_receive_time = real_t(network_watcher.get_deviation(avg_receive_time)) / 1000.0;
+	real_t packet_arrived_in = 0.0;
+	if (now > epoch_received_timestamp) {
+		packet_arrived_in = static_cast<real_t>(now - epoch_received_timestamp) / 1000.0;
+		const real_t delay = packet_arrived_in - next_epoch_expected_in;
+		network_watcher.push(ABS(delay));
+	}
+	epoch_received_timestamp = now;
+	next_epoch_expected_in = next_sync_time;
 
+	const real_t avg_delay = network_watcher.average();
 	// The network quality can be established just by checking the standard
 	// deviation. Stable connections have standard deviation that tend to 0.
-	const real_t net_poorness = MIN(
-			deviation_receive_time / net_sentitivity,
-			1.0);
+	//const real_t deviation_delay = network_watcher.get_deviation(avg_delay);
 
-	const int optimal_frame_delay = Math::lerp(
+	const real_t net_poorness = MIN(1.0, avg_delay / node->get_doll_net_sensitivity());
+
+	const int64_t target_virtual_delay = Math::lerp(
 			node->get_doll_min_frames_delay(),
 			node->get_doll_max_frames_delay(),
 			net_poorness);
 
-	// TODO cache this?
-	const double frames_per_batch = node->get_doll_epoch_batch_sync_rate() * real_t(Engine::get_singleton()->get_iterations_per_second());
-	const double next_batch_arrives_in = Math::ceil(double(next_collect_rate) / frames_per_batch) * frames_per_batch;
+	const real_t epochs_span = target_virtual_delay - current_virtual_delay;
+	const real_t frame_time = 1.0 / real_t(Engine::get_singleton()->get_iterations_per_second());
```

Comment:
Those are equivalent, but I preferred to do it in this way because I used the same formula on the scene sync .cpp file: and it will be easy for me to refactor once I'll detach the net sync processing with physics processing.